.github/workflows/zulip-tests.yml

```diff
@@ -0,0 +1,36 @@
+name: Zulip Tests
+
+on: [pull_request, push]
```

Comment:
This should be
```yml
on:
  push:
    branches:
    - master
  pull_request:
```

otherwise it would be wasteful.

Reply:
I added this because I thought we might want the tests for all the branches.

Reply:
We currently only have a `master` branch, and I'd prefer to default to running tests on new branches, rather than not.

Reply:
The GA workflow will be replicated across all the forks. So in `on: [push, pull_request]` whenever someone pushes a new branch in their own fork, the build will be triggered in their forks. This is not Travis' behavior since CI is by default turned off, but it'd be wasteful energy consumption to the planet, even though it is MSFT's money that is being burned.

Reply:
Aha!  Thanks for teaching me that!  Given that, this makes sense to keep as-is.

Reply:
To clarify, with the version in @rht 's first comment above, would this mean Actions would not be run in users' own forks, unless their PRs happen to be on the master branch? I get the feeling this is something that devs find useful before opening PRs, though enabling the setting to build forks separately via Travis could be something easily missed for starting devs.

What about something like: https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662/10 (or similar)

Reply:
> What about something like: https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662/10 (or similar)
```yml
  push:
  pull_request:
    branches:
    - '**:**'
```
seems to exclude PRs created with non-fork branches, which is not what we want? We want to test every PR, regardless of the source repo.

Reply:
@rht Agreed. This does solve a problem with duplication, but not what we want here.

If I understand this correctly I'm not sure there's a way around this without some way of removing duplicated work between forks (for PRs + fork-branches). However, it would be useful for users to have the ability to enable Actions on their own forks not just for master branches (like with Travis) - without adding a custom commit to their local trees!

Reply:
I would suggest it to allow on all push and PR as the rest of the Zulip does it? Let's not think much about duplicate builds because we would obviously not want to remove user privileges unnecessarily which we provided earlier, just because we want to increase efficiency that too in an unconstrained environment.

Reply:
> However, it would be useful for users to have the ability to enable Actions on their own forks not just for master branches (like with Travis) - without adding a custom commit to their local trees!

When a dev working on their fork pushes a WIP to a branch, it is not always the case that they want their commits to be tested right away.
My solution to this: if a dev explicitly wants to CI-test the fork-branch once they are confident that the code has stabilized, they can just make a PR to the upstream and label it WIP. If this is not feasible, a small commit to `on: [push, pull_request` can be added easily. tl;dr: think of all the wasted builds if you do `on: [push, pull_request` by default.

Reply:
See e.g. how Mypy does it: https://github.com/python/mypy/blob/master/.github/workflows/test.yml.

Reply:
This is a good start then, But we need to update the doc to make people aware of this.

Reply:
Mypy is not the only major Python-based FOSS project restricting CI build on push to master. Flask does it too: https://github.com/pallets/flask/blob/master/.github/workflows/tests.yaml#L2-L10. Several others either use CircleCI or their own CI (e.g. Google), so it's a bit hard to find more examples. My point is that please consider the wasted resource.

Reply:
Great, I do appreciate your thought on resource utilization. Most people consider resources as free or paid and only consider the way it is utilized if there is some restriction or payment involved, but thanks for making me realized that we are developing for the future and it's our duty not to misuse what we get for free or probably it won't last forever. Thanks a lot, @rht.

Reply:
@rht, @neiljp can we continue with this.